### ISSUE TITLE
Simplify HTTP semantics of API, remove PATCH, add etags, conditional  operations 

### ISSUE THREAD

**Issue by zootalures** [fnproject/fn#988]
1) Kill PATCH 
This is for discussion and could be broken down later: 

The API currently uses PATCH  for applying partial modifications to apps and routes  for some subset of operations. 

Patch is problematic as :  
* For some operations it's not suppored (e.g. adding an element to a ist)
* For map modifications you need to elect a null value to delete entries (the semantics are odd and easy to get right) 
* In most cases the client needs the full model to determine what to patch anyway so the patch has exactly the same network trips as a PUT 
* It adds two independent  mutation operations on the server double your code, double your bugs. 

Patch can be safely removed for all resource and replaced with PUT, becomes more powerful when combined with etags and conditional operators below. 

2) Add Etags: (and possibly resource versions) 

Clients have no easy way to detect if a resource  has changed without doing a full diff, internally we have several cases where code is needs to know if a resource is changed but doesn't care how it has changed (e.g. updating containers in the runner) 

Internaly we  should store a  monotonically increasing resource version  for each resource (e.g. as counter or high-resolution timestamp + random) and increment this on change ,  this can be exposed in the body of resources (for easy consumption in code) and also be exposed as a conventional Etag for HTTP operations. 

3) Add HTTP  conditional semantics to operations (GET,PUT)  
As a client I want to know if my resource made the change that I intended - including if somebody else changed the resource between my having read it and written it. 

add support for at least `If-Match` and possibly other HTTP conditionals to allow clients to use optimistic locking for changes - i.e. if the resource wasn't the one I was expecting to change, fail with a 412 and let the client re-cycle the change and try again.

**Comment by rdallman** [comment]
good idea. into it, one explicit goal of #957 was to remove POST too for similar reasons (2 ways to create, and inconsistent across resources). I'm game to only have `GET, PUT, DELETE` for every resource. PATCH and PUT are really similar now, as stated.

> etags

we do have `updated_at` on many resources, this seems like it would satisfy the conditions required for conditional semantics of doing a GET-PUT if we could `if-match` on this? I would like to have more robust version history for certain resources (functions, at least, and probably triggers too), where the version history contains the full objects of each version (for rollbacks & deploys, auditing) - this is related to the etag since that would likely be a server generated monotonic number as well, seems like we could add a `version` field before having a more robust version history as outlined and then use it for that purpose, too.

**Comment by rdallman** [comment]
we're doing this as part of #1038 - likely as one unit of work across all resources, to add revisions to resources after we've added all the resource models as well as the HTTP semantics to if-match on them. new resources will not have PATCH, at which point apps are the only resource with PATCH, and a good time to remove app PATCH may be after trig/func & execution are in place, at the same time that we can remove routes (open to ideas).
